cilium/CFP-37315-bgp-auto-discovery.md

```diff
@@ -0,0 +1,176 @@
+# CFP-37315
+
+**SIG: SIG-BGP**
+
+**Begin Design Discussion:** 2024-02-24
+
+**Cilium Release:** X.XX
+
+**Authors:** Naveen Achyuta <naveen.achyuta22@gmail.com>
+
+## Summary
+
+Cilium’s current BGP implementation requires specifying the peer IP address in the BGP cluster configuration. In large-scale environments with thousands of Kubernetes nodes, managing distinct BGP configuration files (one per rack) becomes impractical.
+
+This proposal offers a simpler alternative: rely on the default gateway of ToR switches to automatically create BGP sessions when no peer IP is provided in the BGP cluster configuration.
+
+## Motivation
+
+Large networks with thousands of Kubernetes nodes cannot feasibly manage the peer IP addresses of all ToR switches within numerous BGP cluster configurations. Allowing Cilium to automatically discover and use the  default gateway as the peer simplifies BGP session creation, reducing operational overhead and configuration complexity.
+
+## Goals
+
+### Auto-Discovery of Peer IPs
+Allow Cilium to establish BGP sessions with the default gateway on the node.
+### Single Config File for All Nodes
+Enable large networks to use a generic BGP configuration file without needing to specify unique peer IP addresses on a per-rack basis.
+
+## Non-Goals
+
+- Allow discovering the peer from arbitrary prefix ([ref](https://github.com/cilium/design-cfps/pull/68#issuecomment-2722443710))
+
+## Proposal
+
+### Overview
+
+At a high level, this proposal modifies how Cilium handles BGP peer IP configuration. If a user omits the peer IP for a BGP process, the Cilium agent attempts to retrieve the node’s default gateway from statedb (which stores local routing information). Cilium can then use this default gateway as the peer IP to establish the BGP session. If the default gateway is not found, the agent skips the neighbor and logs an error for it.
+
+### BGP Cluster Config
+
+If the user enables auto-discovery for a peer, cilium agent should use the discovery mode to get the peer ip. If default gateway mode is being used,
+it should rely on statedb to get default gateway. If not found, it should skip the neighbor and log an error.
+
+Every peer will have to specify the address family under autoDiscovery config block so that we fetch the default gateway for that address family only.
+
+If there are multiple default gateways, select the default gateway with the minimum metric. If it couldn't break the tie in this way, the reconciliation fails.
+
+config proposal:
+```
+bgpInstances:
+  - name: "65001"
+    localASN: 65001
+    peers:
+    - name: "peer1 - 65000"
+      peerASN: 65000
+      peerConfigRef:
+        name: "cilium-peer"
+      autoDiscovery:
+        mode: default-gateway
+        addressFamily: ipv4 // or ipv6
```

Comment:
Sorry, one last suggestion. Probably, we should isolate `addressFamily` field to the isolated stanza. Otherwise, it will be confusing when we introduce another auto discovery mode and the address family doesn't make sense for it.

```suggestion
      defaultGateway:
        addressFamily: ipv4 // or ipv6
```

Reply:
hmm thats a good point. agreed!
if we decide to use bgp unnumbered some day, then that config will not be useful.